### PR TITLE
`factoryFor` as a public API replacing `_lookupFactory`

### PR DESCRIPTION
**Pull request by mixonic**
[Rendered](https://github.com/mixonic/rfcs/blob/factoryFor/text/0000-factoryFor.md)

Based on [conversation from the F2F](https://github.com/emberjs/core-notes/blob/master/ember.js/2016-06/june-11-12_F2F.md#removing-double-extend) and with preliminary review and modifications by @stefanpenner, @rwjblue.

Closes https://github.com/emberjs/rfcs/issues/125

### CODE REVIEW

**Comment by krisselden** [comment]
Let's please use constructor instead and say it is optional, I think it is ok to register a factory to create an object literal or use Object.create().

**Comment by stefanpenner** [comment]
> Let's please use constructor instead and say it is optional, I think it is ok to register a factory to create an object literal or use Object.create().

constructor may have meaning in that position, which is why class was choosen.

**Comment by mixonic** [comment]
@krisselden regarding registration- You're looking for language that it is acceptable to call `owner.register('foo:baz', {create(opts) { return whatever }})`, and that would work with both `lookup` and `factoryFor`? Anything with a `create` method is considered a valid factory?

**Comment by rwjblue** [comment]
@mixonic - That protocol (having a `.create` and possibly an `.extend` method on an object) generally works today, and I think it should be documented (though I'm unsure that documentation effort is specifically related to this RFC).

**Comment by mixonic** [comment]
Seem closely enough related to fold in. Today `extend` is required for DI to be supported, and it is indeed a meaningful change that only `create` will be required when this lands.

**Comment by martndemus** [comment]
I think the how we teach this section should also mention a migration guide for most of the cases that will break.

@rwjblue's example implementation in #125 also seems like it could be added to the RFC. His comment tells the story much better than this 280 line RFC.

Could you also add some reduced examples to the drawbacks section? The links towards projects do not always paint the picture of what is meant very clearly.

**Comment by mixonic** [comment]
@martndemus a migration guide draft seems appropriate here.

> His comment tells the story much better than this 280 line RFC.

😭 which you've asked me to add further details to :trollface: I'll add a link to his comment somewhere relevant.

> Could you also add some reduced examples to the drawbacks section?

There are really not any good use-cases where you need this behavior, so any reduced example would also be pretty contrived. I will attempt to explain what the examples I've linked to are.

**Comment by martndemus** [comment]
The more ELI5 this RFC is, the more easier it is to crowdsource docs/guides updates. 😄

**Comment by mixonic** [comment]
I've pushed a round of edits here, much improved. Please give it another look!

File: text/0000-factoryFor.md [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+Any state stored on the constructor would of course leak between the tests.
+
+An example implementation of `factoryFor` can be reviewed [on this GitHub
+comment](https://github.com/emberjs/rfcs/issues/125#issuecomment-193827658).
+
+##### Implications for `owner.register`
+
+Currently, factories registered into Ember's DI system are required to
+provide an `extend` method. Removing support for extend-based DI in `_lookupFactory`
+will permit factories without `extend` to be registered. Instead factories
```
Comment (stefanpenner) [inline text/0000-factoryFor.md]: I agree with this change, but it is likely the second "breaking change" in this RFC. Is this a reasonable concern, or is it ok, and if so how do we migrate forwards + minimize any potential fallout.

**Comment by stefanpenner** [comment]
I am slightly confused, what part of the rollout do we actually diverge the inheritance structure to the new simpler form?

**Comment by mixonic** [comment]
@stefanpenner this proposal is fairly aggressive in that it suggests removal of the `_lookupFactory` API in Ember 2.9. See: https://github.com/mixonic/rfcs/blob/eccbd06ef6dfd4b50d1c8d4bb45ad58d110334f4/text/0000-factoryFor.md#step-2-deprecate-applicationinstance_lookupfactory-in-ember-28-remove-in-29 Ember 2.8-LTS would provide both APIs, giving LTS supporting addons/apps until Ember 2.12-LTS to make the migration.

At the same time, this RFC also proposes releasing a polyfill for `factoryFor` going back to at least Ember 2.4. An application (or addon) migration strategy would be to migrate to the new API via the polyfill, then remove that polyfill when support for Ember pre-2.9 is no longer a requirement.

**Comment by stefanpenner** [comment]
@mixonic this RFC describe the API, but not when we deprecate the internal behavior.

**Comment by stefanpenner** [comment]
In the meeting we also discussed wrapping `.class` in a proxy for development, a proxy which:

* disallows all sets & obvious mutations.
* disallows `reopen` `reopenClass`
* allows all reads
* allows all other operations

**Comment by stefanpenner** [comment]
@mixonic  / @rwjblue I (well actually @krisselden) have found a legit in-the-wild example of how we will end up breaking real code by no longer calling `extend`: https://github.com/emberjs/ember.js/blob/master/packages/ember-glimmer/lib/views/outlet.js#L49

One work-around would be to only not call `extend` if it is the `extend` we know to be the one on an ember object...

**Comment by krisselden** [comment]
I would not be surprised if there is not addons that use extend to return a custom factory

**Comment by stefanpenner** [comment]
> I would not be surprised if there is not addons that use extend to return a custom factory

Yes, i believe we can detect those cases and do the old thing.

**Comment by sukima** [comment]
To help this RFC along, our team is desperately trying to migrate from 1.13. One of the hardest blockers of this is our app design relied heavily on JS only `imports` and window/Ember globals. Newer implementations of these features use services now which are unavailable without a container. To help this migration we would love to get our dependency injections set for our `Ember.Object` based models so that when we do eventually convert to ember-data the services we use would already be available.

The downsides are:

* The docs show `_lookupFactory` as **private**. This makes our team members nervous.
* The method starts with an underscore which makes our team doubly nervous.

Although `_lookupFactory` is a de facto standard and used among addon authors in the wild it is still an up hill battle in some teams to convince them that a private method that starts with an underscore is technically *safe*.

This RFC would offer a sanctioned way to create objects via a factory with dependency injection where the usual `Ember.Object.create` would not otherwise. The examples in this RFC are an exact match for what we need in our app to move it forward. For now we will use `_lookupFactory` but would really love to see this RFC move into a recognized implementation and a polyfill to get older apps there.

ember-data is great that it can do this stuff automagically but it sure is nice to be able to do these tricks on our own. As an addon author it opens up a large opportunity for some very powerful features. I hope this testimony of one possible use case in a production app helps adds some context to this RFC.

**Comment by sukima** [comment]
For what it's worth here is how we worked with this (with out the ember-getowner-polyfill):

```javascript
import Ember from 'ember';

const { isPresent, Service, deprecate } = Ember;

function getOwner(ctx) {
  deprecate(
    'Ember.getOwner() is available now; please cleanup references to this.container',
    isPresent(Ember.getOwner),
    {id: 'container-lookup'}
  );
  return isPresent(Ember.getOwner) ? Ember.getOwner(ctx) : ctx.container;
}

export default Service.extend({
  createRecord(modelName, payload) {
    const owner = getOwner(this);
    const lookupFactory = owner._lookupFactory || owner.lookupFactory;
    const ModelFactory = lookupFactory.call(owner, `model:${modelName}`);
    return ModelFactory.create(payload);
  }
});
```

**Comment by mixonic** [comment]
This RFC is basically in FCP pending the process item of a tweet from https://twitter.com/emberjs and tag. If you've any remaining concerns you believe are missed here, please let me know!

**Comment by pixelhandler** [comment]
@mixonic I use  `_lookupFactory` in my data persistence addon. Today I read over the RFC and am very pleased with the thought and depth put into this RFC.

As an addon developer I think it will be straight forward to make the upgrade to `factoryFor` and at the same time continue support `_lookupFactory` as well. (the build for the addon tests v1.13 though canary).

👍

**Comment by chadhietala** [comment]
I would like to propose we don't inject the deprecated container into the instances returned from `factoryFor` and we mitigate that to the deprecated `_lookupFactory` path.

**Comment by mixonic** [comment]
@locks thats for the tagging!

@chadhietala Can you say more? There is no "deprecated container" described here.

**Comment by dgeb** [comment]
*Note: To make the Final Comment Period official, we need to announce it here and on Twitter, so we're restarting the period now.*

---

This RFC is now entering a ["final comment period"](https://github.com/emberjs/rfcs/pull/149). This period is intended to focus the community's attention for a final round of consideration and feedback before an RFC is merged.

Please carefully consider the implications of this RFC and how it will affect your Ember projects. Now would be a good time to re-raise any concerns that you feel haven't been fully addressed.

If no significant blockers are raised by **October 7**, we plan to merge this RFC and make plans for its implementation. :shipit:

**Comment by rwjblue** [comment]
Looks like no additional comments came through during the FCP process.  Lets do it!!!

**Comment by pixelhandler** [comment]
@rwjblue should the "Final Comment Period" label be removed?

**Comment by dgeb** [comment]
@pixelhandler good call - done.

**Comment by chadhietala** [comment]
See https://github.com/emberjs/ember.js/pull/14360 for the WIP implementation.
